### ISSUE TITLE
Compilation with -fmodules-ts and use inside of a module

### ISSUE THREAD

**Issue by LeHibou2** [nlohmann/json#3472]
### Description

Hi,
When just including #include <nlohmann/json.hpp> in a module, the compiler complains about impossible references.
It is not happening when not inside a module.

### Reproduction steps

Just include the header inside a module and compile with -fmodules-ts.

This is the triggering flag, really.

### Expected vs. actual results

No error

### Minimal code example

```Shell
//mod.cxx:

module;
#include <iostream>
#include <string_view>
#include <string>
#include <nlohmann/json.hpp>
export module hello;

export {
        namespace X {

                void greeter ( std::string_view const &name ){
                        std::string h = "exported greeter";

                        std::cout << "Hello " << name << "!" << h << "\n";

                }

        }
}


// main.cpp
import hello;

int main ( void ){

        X::greeter("ok");
        return 0;
}

// compile with
~/gcc-12.1/g++ -fmodules-ts -std=c++20 -I/my_path/include/ mod.cxx main.cpp
```

### Error messages

```Shell
`/usr/include/nlohmann/adl_serializer.hpp:41:17: error: ‘static decltype ((nlohmann::{anonymous}::to_json(j, forward<ValueType>(val)), void())) nlohmann::adl_serializer<T, SFINAE>::to_json(BasicJsonType&, ValueType&&) [with BasicJsonType = nlohmann::basic_json<>; ValueType = bool&; <template-parameter-1-1> = bool; <template-parameter-1-2> = void; decltype ((nlohmann::{anonymous}::to_json(j, forward<ValueType>(val)), void())) = void]’ references internal linkage entity ‘constexpr const nlohmann::detail::to_json_fn& nlohmann::{anonymous}::to_json’
   41 |     static auto to_json(BasicJsonType& j, ValueType&& val) noexcept(`
```

### Compiler and operating system

Linux GCC-12.1

### Library version

Latest

### Validation

- [ ] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

**Comment by falbrechtskirchinger** [comment]
If the issue is what I suspect it is, this should be easily fixable. I'll take a look.

**Comment by falbrechtskirchinger** [comment]
I was able to fix the linkage issues by marking the relevant entities `inline` but your code then results in an ICE:
```
 $ g++-13.0.0 -fmodules-ts -std=c++20 -I../fix-c++20-issues-part2/include/ mod.cxx main.cpp
main.cpp: In function ‘int main()’:
main.cpp:4:12: internal compiler error: Segmentation fault
    4 |         X::greeter("ok");
      |            ^~~~~~~
0x7fcf08bf793f ???
        /var/tmp/portage/sys-libs/glibc-2.35-r4/work/glibc-2.35/signal/../sysdeps/unix/sysv/linux/x86_64/libc_sigaction.c:0
0x7fcf08be3389 __libc_start_call_main
        ../sysdeps/nptl/libc_start_call_main.h:58
0x7fcf08be343b __libc_start_main_impl
        ../csu/libc-start.c:392
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://bugs.gentoo.org/> for instructions.
```
(And GCC 12 crashes as well.)

You can grab the fixed code from my `modules` branch. Maybe you can figure out why GCC is crashing.

https://github.com/falbrechtskirchinger/json/tree/modules

(The code is running through CI right now. If there're any problems I will update this post.)

**Comment by LeHibou2** [comment]
Thanks!
There is a much deeper issue in the handling of modules in gcc and I reported a bug this morning.

I will try your fix once I get some lights 👍👍

**Comment by falbrechtskirchinger** [comment]
@nlohmann This bug can be closed. Fixed by #3446 (4f744f4).
